### PR TITLE
Refactored comms to add JupyterLab support

### PR DESCRIPTION
**Pull request by jlstevens**
This PR is still WIP and there are a lot of issues to address:

- [x] Made comms code consistent with holoviews.
- [x] Works in classic notebook
- [x] Displays widgets in jupyterlab
- [x] Updates parameter values in jupyterlab
- [x] All functionality relating to the bokeh comms and plots option.
- [x] Currently removing duplication of comms code by importing from holoviews. HoloViews should not be a parambokeh dependency but we do need a comms package for pyviz.
- [x] Currently needs a ``HoloViews`` object to be introduced in JS which isn't appropriate for other pyviz projects.

Right now the problem I am trying to address is to get comms working in jupyterlab as the parameter values are not updating.

Note that this requires the pyviz jupyterlab extension. I have the bokeh jupyterlab extension but I'm not yet sure whether or not it is actually required.

### CODE REVIEW

File: parambokeh/__init__.py [hunk 1]
```diff
@@ -359,11 +388,16 @@ def _get_customjs(self, change, p_name):
...
         data_template = "data = {{p_name: '{p_name}', value: cb_obj['{change}']}};"
+        proxy_object = HOLOVIEWS_PROXY
         fetch_data = data_template.format(change=change, p_name=p_name)
         self_callback = JS_CALLBACK.format(comm_id=self.comm.id,
                                            timeout=self.timeout,
-                                           debounce=self.debounce)
-        js_callback = CustomJS(code=fetch_data+self_callback)
+                                           debounce=self.debounce,
+                                           plot_id=self.comm.id)
+        js_callback = CustomJS(code='\n'.join([HOLOVIEWS_PROXY,
```
Comment (philippjfr) [inline parambokeh/__init__.py]: Putting the proxy code in here will reset the comms and kernels registries after they have been set up. At minimum you'd have to ensure that it does not override them if they are already defined.

Comment (jlstevens) [reply parambokeh/__init__.py]: Ok, I'll give that a go.

File: parambokeh/__init__.py [hunk 2]
```diff
@@ -29,6 +35,29 @@
...
+def notebook_show(obj, doc, target):
+    """
+    Displays bokeh output inside a notebook and returns a CommsHandle.
+    """
+    mime = 'application/vnd.holoviews_exec.v0+json'
+    bokehcomm = bokeh.io.notebook.get_comms(target)
+    bokeh_script, bokeh_div, _ = bokeh.embed.notebook.notebook_content(obj, target)
+    publish_display_data(data={'text/html': encode_utf8(bokeh_div)})
+    publish_display_data(data={mime: '', 'application/javascript': bokeh_script},
+                         metadata={mime: {'id': target}})
```
Comment (philippjfr) [inline parambokeh/__init__.py]: This needs to be the ``self.comm.id`` rather than the target.

Comment (jlstevens) [reply parambokeh/__init__.py]: Thanks. This is what was needed to fix it.

**Comment by jlstevens** [comment]
The basics now work in both classic notebook and jupyterlab: you can display some widgets, manipulate them and use that to update parameters.

I'll now look into the other uses of comms (e.g relating to ``View`` parameters/``plots``) but after that, the remaining decisions involve changes across repos (making a comms package, renaming the ``HoloViews`` object etc).

**Comment by jlstevens** [comment]
Here is a related pyviz issue: https://github.com/pyviz/jupyterlab_pyviz/issues/4

It isn't the whole story as the comms code itself references ``HoloViews`` and that is where it currently lives, though it should be pulled out and made general to pyviz.

**Comment by jlstevens** [comment]
With the commit by @philippjfr, I believe the TODO items have been ticked off except those that require more discussion across pyviz.

**Comment by jlstevens** [comment]
Updated the imports to use the new [``pyviz_comms``](https://github.com/pyviz/pyviz_comms/pull/5) package.

**Comment by jlstevens** [comment]
Tests are now green and I have ticked off the TODO items above. It isn't perfect but I think it is ready to review and merge before improving upon in subsequent PRs. The ``pyviz_comms`` package is now required and available on the pyviz conda channel. This will be available on PyPI soon.

**Comment by philippjfr** [comment]
Looks good.

**Pull request merged**